src/spi.rs

```diff
@@ -300,10 +308,11 @@ pub mod config {
 }
 
 pub struct SpiBusDriver<T> {
-    _lock: Lock,
+    lock: Option<Lock>,
```

Comment:
Yeah. Think about multiple devices on a single bus. This case is impossible to do asynchronously (assuming no changes are made to the ESP IDF itself and no horrible hacks). Reason: `lock the bus` is **blocking/synchronous** and unfortunately we have to use so as to implement proper `embedded-hal` (or if you wish - "Rust") transactions, as the native ESP IDf transactions are constrained in terms of buffer size. Therefore:
* Either you have just **one** device on the bus, and then locking the bus always completes without waiting
* Or you need to lift the "lock the bus" requirement - which - of course - means you are not doing "real" transactions, from e-hal POV, but at least the stuff runs without blocking.

The `esp-idf-hal` SPI public `*_async` APIs do take a `lock_bus: bool` extra parameter precisely for that reason. I mean, the user - in the current state of things, has to pick the lesser evil. Which one is the lesser evil (try to lock the bus and block if there are multiple devices on it or don't lock the bus but then no real transactions - no big deal for stuff like LCD I guess) *only the user knows*.

Reply:
Ah, I hadn't considered the blocking aspect of this.
I see your point. I think it'd be better if users just make the individual calls to `read`, `write`, etc. if they what `lock_bus = false` behaviour.

`spi_device_acquire_bus` takes a timeout though, so you could make it poll like you've done with `spi_device_get_trans_result` I think.

Reply:
> `spi_device_acquire_bus` takes a timeout though, so you could make it poll like you've done with `spi_device_get_trans_result` I think.

Timeout is of no use if I don't get notified when to try again. Hmmm...

Reply:
Well, don't you have that problem already with `spi_device_queue_trans`? It internally locks the bus, which is why it takes the timeout parameter as well.

```Rust
                loop {
                    match esp!(unsafe {
                        spi_device_queue_trans(handle, &mut transaction as *mut _, delay::NON_BLOCK)
                    }) {
                        Ok(_) => break,
                        Err(e) if e.code() != ESP_ERR_TIMEOUT => return Err(e),
                        _ => NOTIFIER[*host as usize].wait().await,
                    }
                }
```

Ahhh the shared notifier makes this work. Pesky pesky. Idk if this is too far but maybe some shared state per driver to keep a bag of all wakers waiting for transactions to end maybe?

There's also the option of calling the waker immediately but I'll get scolded for that suggestion haha.

Reply:
But you don't **know** when the bus lock is released, because you are not notified about that - we always get down to this issue.

"Calling the waker directly" is essentially busy-looping which would eat up ALL CPU cycles of the executor scheduled on that thread and all other tasks assigned to the same executor would starve. At the very least, the loop should have some sort of back-off timeout strategy which is based on awaiting an embassy timer or an esp idf async timer to trigger, but that might introduce unacceptable delays in the SPI send/receive code. Still, a relatively simple solution.

Another approach would be a "gargantuan" async+blocking lock of sorts around the native spi bus lock that uses its own locking mechanism (and a queue of wakers!) to ensure that when it tries to lock the bus lock, nobody else is competing for *that* lock as well. Very complex and won't work when you have both native esp idf drivers on the same spi bus (think the ethernet spi drivers) with your rust device drivers, as the former would be completely unaware of your "gargantuan" lock.

The issue is so comple,, that I think we are better off thinking about it after the first release of these drivers.

Reply:
Yes I agree with you. Now that you've pointed out the native esp-idf driver problem, the issue I brought up still stands.

How do you solve this problem for `spi_device_queue_trans`? Since you don't know when the bus lock is released. (`spi_device_queue_trans` and `spi_device_polling_transmit` locks the bus)

Reply:
We pre-lock it. Which we anyway have to do if the `keep-cs-active` flag is raised (and it always is by us, as the ESP IDF SPI transactions don't fit the notion of E-HAL transactions). Otherwise the SPI driver returns an error.

Reply:
... but as the table at the top of the driver says - you can't currently use async when you have multiple devices on the bus. Or actually you can, but it will not be async.

Reply:
Ah fair enough, it all makes sense now.
At least folks who want multiple devices can create an `SpiBusDriver`, share it with an async mutex and use software CS.
In a future, it'd be interesting to consider injecting some kinda thread to perform the blocking.

A fun thing you can do now to enforce the documentation is to only implement the async methods on `SpiDevice` when `T` is `BorrowMut` instead of just `Borrow`.